### ISSUE TITLE
[stable/redis-ha] allow running haproxy as DaemonSet

### ISSUE THREAD

**Issue by viktortnk** [helm/charts#17717]
**Is your feature request related to a problem? Please describe.**
I would like to be able to direct traffic from application pods to haproxy on the same machine to reduce latency on potentially going to other VM

**Describe the solution you'd like**
Allowing to run haproxy as DaemonSet, so that it is possible to access Redis via the same node

**Additional context**
Similar to this setup
![image](https://user-images.githubusercontent.com/320239/66257992-bab4cd80-e797-11e9-8ead-40ee89eb813f.png)

**Comment by DandyDeveloper** [comment]
This would need to be well thought out, it's also a little bit painful as the implication of a DS means it will run on every node (assuming you don't use taints/node selector) which could be a headache for people using this. 

I'm also wondering how bad the latency would be between VMs to warrant this. The TCP requests travelling between the machines is tiny, unless you have absolutely massive key values. 

Also, doesn't this defeat the purpose of HA? Surely if you run 1 HAProxy and want that traffic routed to the pod running on the same VM, what happens if you lose the pod?

How would this also work in the case where you have the master/slave hierarchy? Redis will determine the master based on the available hosts and the quorum. You can't force the master to be the same pod running on the same machine as the HAProxy.  

@ssalaues I'd want your opinion on this. I get the use case, but it seems like a lot of work that may actually cause more headache for people in the long run.

**Comment by ssalaues** [comment]
I can definitely understand the need for this as well (large scale, hpc, etc) where milliseconds really count. However there are some roadblocks that I can see, implementation issues aside.

- How would you guarantee that a traffic coming into a HAproxy DaemonSet pod on one node doesn't get routed to a pod on another node? Since the `redis-ha-haproxy` service (at the Kubernetes level) regardless of controller type will do it's own internal load balancing to service endpoints which in this case would be the Haproxy DaemonSet pods. This can be changed on your specific K8s install from the default of `round-robin` to something like `shortest expected delay` but still doesn't guarantee traffic stays on the same node and even less so configurable from the chart level.
https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies

- I think something similar to what you're looking for could be achieved with some modifications to the chart. If you are trying to scale writes out (and possibly reads) supporting some level of redis cluster mode through the chart can be possible. If you're trying to only scale reads then the cluster mode isn't necessary but the general idea is about the same. Although I haven't tested it but here's the general idea:
![redis](https://user-images.githubusercontent.com/24439848/66959491-55769d00-f01f-11e9-8462-d6eaf802fa06.png)
Basically have each release tainted/nodeSelector/Affinity/etc (a soft Affinity might be best to allow for rescheduling in case of a down node but YMMV) to a single Node but having the pods in the 3 nodes a part of the same redis cluster.

There are obviously still many things left to consider here (HA especially) and I agree with @DandyDeveloper that this implementation really has to be well thought out to avoid making the chart more difficult for others to use. Overall however, I think staying away from the DaemonSets is ideal especially since I don't see how it would help in this use-case.
